### PR TITLE
8343629: More MergeStore benchmark

### PR DESCRIPTION
**Pull request by wenshao**
1. Added the putBytes4 benchmark, which corresponds to StringBuilder appendNull
2. Optimized the putChars4/setInt/setLong series of benchmarks to reduce extra overhead and more accurately reflect performance differences.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343629](https://bugs.openjdk.org/browse/JDK-8343629): More MergeStore benchmark (**Enhancement** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21659/head:pull/21659` \
`$ git checkout pull/21659`

Update a local copy of the PR: \
`$ git checkout pull/21659` \
`$ git pull https://git.openjdk.org/jdk.git pull/21659/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21659`

View PR using the GUI difftool: \
`$ git pr show -t 21659`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21659.diff">https://git.openjdk.org/jdk/pull/21659.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21659#issuecomment-2457519335)
</details>

### CODE REVIEW

**Comment by wenshao** [comment]
@eme64 

Below are the performance numbers running under AMD EPYC™ Genoa (x64), where the scenario of putBytes4GetBytes is
```java
"null".getBytes(0, 4, bytes4, off);
```
Is it possible to do MergeStore in this scenario?


```
Benchmark                          Mode  Cnt      Score       Error  Units
MergeStoreBench.getCharB           avgt    5   6038.532 ±   533.982  ns/op
MergeStoreBench.getCharBU          avgt    5   4923.182 ±   163.872  ns/op
MergeStoreBench.getCharBV          avgt    5   3111.268 ±    84.077  ns/op
MergeStoreBench.getCharC           avgt    5   2245.270 ±    33.559  ns/op
MergeStoreBench.getCharL           avgt    5   6109.519 ±   249.512  ns/op
MergeStoreBench.getCharLU          avgt    5   4552.425 ±   161.933  ns/op
MergeStoreBench.getCharLV          avgt    5   2239.866 ±    91.853  ns/op
MergeStoreBench.getIntB            avgt    5   8163.035 ±   137.565  ns/op
MergeStoreBench.getIntBU           avgt    5   9136.199 ±   259.491  ns/op
MergeStoreBench.getIntBV           avgt    5    314.123 ±     4.510  ns/op
MergeStoreBench.getIntL            avgt    5   7879.011 ±    10.759  ns/op
MergeStoreBench.getIntLU           avgt    5   8968.715 ±   268.414  ns/op
MergeStoreBench.getIntLV           avgt    5   2228.228 ±     1.510  ns/op
MergeStoreBench.getIntRB           avgt    5   8618.141 ±    22.545  ns/op
MergeStoreBench.getIntRBU          avgt    5  11239.977 ±   447.754  ns/op
MergeStoreBench.getIntRL           avgt    5   9060.754 ±   236.147  ns/op
MergeStoreBench.getIntRLU          avgt    5   9365.050 ±   154.357  ns/op
MergeStoreBench.getIntRU           avgt    5   2540.704 ±    75.198  ns/op
MergeStoreBench.getIntU            avgt    5   2508.954 ±    74.999  ns/op
MergeStoreBench.getLongB           avgt    5  24940.668 ± 16857.311  ns/op
MergeStoreBench.getLongBU          avgt    5  14126.468 ±   329.241  ns/op
MergeStoreBench.getLongBV          avgt    5    607.128 ±    23.775  ns/op
MergeStoreBench.getLongL           avgt    5  25519.679 ± 15393.727  ns/op
MergeStoreBench.getLongLU          avgt    5  14598.271 ±   481.158  ns/op
MergeStoreBench.getLongLV          avgt    5   2227.659 ±    16.334  ns/op
MergeStoreBench.getLongRB          avgt    5  25158.839 ± 18209.451  ns/op
MergeStoreBench.getLongRBU         avgt    5  14005.082 ±   208.154  ns/op
MergeStoreBench.getLongRL          avgt    5  25303.319 ± 14775.524  ns/op
MergeStoreBench.getLongRLU         avgt    5  14481.847 ±   309.623  ns/op
MergeStoreBench.getLongRU          avgt    5   3065.744 ±    15.405  ns/op
MergeStoreBench.getLongU           avgt    5   3048.522 ±     0.704  ns/op
MergeStoreBench.putBytes4          avgt    5    933.283 ±     6.197  ns/op
MergeStoreBench.putBytes4GetBytes  avgt    5   5917.932 ±   199.901  ns/op
MergeStoreBench.putBytes4U         avgt    5    944.097 ±    25.902  ns/op
MergeStoreBench.putBytes4X         avgt    5    944.714 ±    18.924  ns/op
MergeStoreBench.putChars4B         avgt    5   5679.262 ±   154.030  ns/op
MergeStoreBench.putChars4BU        avgt    5   1143.133 ±     4.250  ns/op
MergeStoreBench.putChars4BV        avgt    5   4530.941 ±   124.318  ns/op
MergeStoreBench.putChars4C         avgt    5   1138.541 ±    27.843  ns/op
MergeStoreBench.putChars4L         avgt    5   5647.885 ±   112.363  ns/op
MergeStoreBench.putChars4LU        avgt    5   1142.501 ±     4.400  ns/op
MergeStoreBench.putChars4LV        avgt    5   1143.770 ±     3.435  ns/op
MergeStoreBench.putChars4S         avgt    5   1141.919 ±    36.528  ns/op
MergeStoreBench.setCharBS          avgt    5   6114.143 ±   144.826  ns/op
MergeStoreBench.setCharBV          avgt    5   3607.599 ±    87.720  ns/op
MergeStoreBench.setCharC           avgt    5   4510.196 ±     5.445  ns/op
MergeStoreBench.setCharLS          avgt    5   5641.424 ±   195.167  ns/op
MergeStoreBench.setCharLV          avgt    5   2267.712 ±    40.752  ns/op
MergeStoreBench.setIntB            avgt    5   8049.368 ±   233.618  ns/op
MergeStoreBench.setIntBU           avgt    5  18052.279 ±  2428.567  ns/op
MergeStoreBench.setIntBV           avgt    5   3287.905 ±    63.375  ns/op
MergeStoreBench.setIntL            avgt    5   2135.887 ±    62.601  ns/op
MergeStoreBench.setIntLU           avgt    5   4795.636 ±    74.974  ns/op
MergeStoreBench.setIntLV           avgt    5   2154.363 ±    81.324  ns/op
MergeStoreBench.setIntRB           avgt    5  13895.941 ±  7981.782  ns/op
MergeStoreBench.setIntRBU          avgt    5  14756.267 ±  1585.571  ns/op
MergeStoreBench.setIntRL           avgt    5   3284.792 ±    37.939  ns/op
MergeStoreBench.setIntRLU          avgt    5   5958.555 ±    27.404  ns/op
MergeStoreBench.setIntRU           avgt    5   5983.119 ±    79.627  ns/op
MergeStoreBench.setIntU            avgt    5   4848.655 ±   168.466  ns/op
MergeStoreBench.setLongB           avgt    5  31871.401 ±  1233.822  ns/op
MergeStoreBench.setLongBU          avgt    5  25704.975 ±  5105.792  ns/op
MergeStoreBench.setLongBV          avgt    5   2199.367 ±    69.511  ns/op
MergeStoreBench.setLongL           avgt    5   5486.926 ±    30.874  ns/op
MergeStoreBench.setLongLU          avgt    5   4503.212 ±    81.635  ns/op
MergeStoreBench.setLongLV          avgt    5   2144.943 ±    38.944  ns/op
MergeStoreBench.setLongRB          avgt    5  30338.353 ±  1631.512  ns/op
MergeStoreBench.setLongRBU         avgt    5  25025.442 ±  2690.138  ns/op
MergeStoreBench.setLongRL          avgt    5   4553.245 ±   128.721  ns/op
MergeStoreBench.setLongRLU         avgt    5   4793.427 ±     1.474  ns/op
MergeStoreBench.setLongRU          avgt    5   4803.963 ±    74.017  ns/op
MergeStoreBench.setLongU           avgt    5   4564.326 ±   146.283  ns/op
```

**Comment by eme64** [comment]
> ```java
> ```java
> "null".getBytes(0, 4, bytes4, off);
> ```
>   
> Is it possible to do MergeStore in this scenario?
> ```

I don't know. What do the logs say? And what does it currently compile down to, i.e. what assembly instructions?

Otherwise I think this update seems reasonable.

**Comment by eme64** [comment]
It would be nice if you could do some summary / explanation: which cases do still not optimize, and why?

**Comment by eme64** [comment]
For that, it would be helpful if you had a run with, and one without `MergeStores` enabled - then we can easily compare the performance!

**Comment by eme64** [comment]
You can find an example of how to do that easily here:
https://github.com/openjdk/jdk/pull/19970/files#diff-9072c369f5b541ef9fca3ad8320aa59e88cc72f203c03da58100b1d111ffc324R746-R749

**Comment by wenshao** [comment]
> > ```java
> > "null".getBytes(0, 4, bytes4, off);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is it possible to do MergeStore in this scenario?
> 
> I don't know. What do the logs say? And what does it currently compile down to, i.e. what assembly instructions?
> 
> Otherwise I think this update seems reasonable.

My thinking is this:

```java
StringBuilder buf = new StringBuilder();
// ...
buf.append("null");
```

The calling path is as follows:
```
AbstractStringBuilder::append -> AbstractStringBuilder::putStringAt -> String::getBytes(byte[], int, byte) -> System::arraycopy
```

In this scenario, if System::arraycopy can be optimized to use putInt or putLong, performance can be improved.

It is similar in the String concatenation scenario
```java
String f(int i) {
	return "abcd" + i;
}
```

Here `StringConcatHelper::prepend(int, byte, byte[], String, String)` is called, and then `String::getBytes(byte[], int, byte) -> System::arraycopy`

```java
package java.lang;
class StringConcatHelper {
    static int prepend(int index, byte coder, byte[] buf, String value, String prefix) {
        index -= value.length();
        if (coder == String.LATIN1) {
            value.getBytes(buf, index, String.LATIN1);
            index -= prefix.length();
            prefix.getBytes(buf, index, String.LATIN1);
        } else {
            value.getBytes(buf, index, String.UTF16);
            index -= prefix.length();
            prefix.getBytes(buf, index, String.UTF16);
        }
        return index;
    }
}
```

Here is similar to the above, can we optimize "abcd".getBytes to putInt or putLong?

In summary, can we optimize the System::arraycopy of a stable byte[] with a length of 4 to putInt?

**Comment by wenshao** [comment]
@eme64 Why is there no noticeable difference in the performance of `+/-MergeStores`

|   | -MergeStores  | +MergeStores | delta |
| --- | --- | --- | --- |
| getCharB | 5900.246 | 5902.316 | -0.04% |
| getCharBU | 4865.881 | 4866.630 | -0.02% |
| getCharBV | 3084.194 | 3078.657 | 0.18% |
| getCharC | 2233.422 | 2232.788 | 0.03% |
| getCharL | 6032.213 | 6028.447 | 0.06% |
| getCharLU | 4492.928 | 4482.773 | 0.23% |
| getCharLV | 2220.004 | 2220.231 | -0.01% |
| getIntB | 7996.907 | 8050.658 | -0.67% |
| getIntBU | 9041.783 | 9035.892 | 0.07% |
| getIntBV | 309.469 | 308.076 | 0.45% |
| getIntL | 7887.687 | 7881.362 | 0.08% |
| getIntLU | 8856.416 | 8863.707 | -0.08% |
| getIntLV | 2225.803 | 2225.789 | 0.00% |
| getIntRB | 8619.974 | 8616.985 | 0.03% |
| getIntRBU | 11098.237 | 11100.091 | -0.02% |
| getIntRL | 8959.808 | 8958.688 | 0.01% |
| getIntRLU | 9237.407 | 9236.465 | 0.01% |
| getIntRU | 2502.967 | 2503.585 | -0.02% |
| getIntU | 2492.784 | 2492.675 | 0.00% |
| getLongB | 24807.583 | 24797.555 | 0.04% |
| getLongBU | 14022.093 | 14008.556 | 0.10% |
| getLongBV | 601.878 | 600.904 | 0.16% |
| getLongL | 25076.552 | 25111.661 | -0.14% |
| getLongLU | 14470.997 | 14474.230 | -0.02% |
| getLongLV | 2223.678 | 2223.882 | -0.01% |
| getLongRB | 24769.555 | 24778.684 | -0.04% |
| getLongRBU | 14017.091 | 14024.421 | -0.05% |
| getLongRL | 25070.811 | 25085.936 | -0.06% |
| getLongRLU | 14462.097 | 14467.410 | -0.04% |
| getLongRU | 3056.826 | 3056.270 | 0.02% |
| getLongU | 3045.057 | 3045.650 | -0.02% |
| putBytes4 | 928.032 | 928.111 | -0.01% |
| putBytes4GetBytes | 5876.794 | 5875.995 | 0.01% |
| putBytes4U | 926.596 | 928.596 | -0.22% |
| putBytes4X | 927.929 | 927.928 | 0.00% |
| putChars4B | 5635.803 | 5635.872 | 0.00% |
| putChars4BU | 1142.948 | 1141.809 | 0.10% |
| putChars4BV | 4482.613 | 4480.597 | 0.04% |
| putChars4C | 1132.133 | 1132.881 | -0.07% |
| putChars4L | 5640.644 | 5632.055 | 0.15% |
| putChars4LU | 1141.009 | 1142.132 | -0.10% |
| putChars4LV | 1133.833 | 1133.137 | 0.06% |
| putChars4S | 1132.469 | 1132.250 | 0.02% |
| setCharBS | 6080.539 | 6081.117 | -0.01% |
| setCharBV | 3598.374 | 3591.190 | 0.20% |
| setCharC | 4497.279 | 4544.706 | -1.04% |
| setCharLS | 5615.475 | 5620.162 | -0.08% |
| setCharLV | 2249.104 | 2245.083 | 0.18% |
| setIntB | 7999.139 | 8030.850 | -0.39% |
| setIntBU | 17922.810 | 17942.929 | -0.11% |
| setIntBV | 3237.265 | 3224.414 | 0.40% |
| setIntL | 2124.492 | 2109.906 | 0.69% |
| setIntLU | 4772.256 | 4801.314 | -0.61% |
| setIntLV | 2110.382 | 2120.022 | -0.45% |
| setIntRB | 13773.518 | 13775.889 | -0.02% |
| setIntRBU | 14752.651 | 14754.926 | -0.02% |
| setIntRL | 3226.597 | 3227.019 | -0.01% |
| setIntRLU | 5862.400 | 5882.564 | -0.34% |
| setIntRU | 5915.139 | 5917.139 | -0.03% |
| setIntU | 4794.627 | 4780.927 | 0.29% |
| setLongB | 31661.626 | 31598.635 | 0.20% |
| setLongBU | 25681.380 | 25622.835 | 0.23% |
| setLongBV | 2167.426 | 2164.900 | 0.12% |
| setLongL | 5380.433 | 5321.645 | 1.10% |
| setLongLU | 4281.526 | 4280.263 | 0.03% |
| setLongLV | 2109.982 | 2110.138 | -0.01% |
| setLongRB | 29807.728 | 29826.089 | -0.06% |
| setLongRBU | 24973.926 | 24903.052 | 0.28% |
| setLongRL | 4518.310 | 4518.594 | -0.01% |
| setLongRLU | 4792.258 | 4795.612 | -0.07% |
| setLongRU | 4796.491 | 4792.139 | 0.09% |
| setLongU | 4280.624 | 4507.839 | -5.04% |

File: test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java [hunk 1]
```diff
@@ -1129,4 +1149,9 @@ public static void putShortLU(byte[] array, int offset, int c) {
         UNSAFE.putByte(array, address    , (byte) (c     ));
         UNSAFE.putByte(array, address + 1, (byte) (c >> 8));
     }
+
+    @Fork(value = 1, jvmArgsPrepend = {
```
Comment (eme64) [inline test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java:1153]:
```suggestion
    @Fork(value = 1, jvmArgs = {
```
Can you make this change, and run the benchmarks again?
There was a recent JMH build script change, and all usages of `jvmArgsPrepend` in JMH tests were supposed to be changed to `jvmArgs`. I think in your case the flag is actually not applied.

Not sure if that is true, but it looks that way to me.

https://github.com/openjdk/jdk/pull/21800

**Comment by eme64** [comment]
> @eme64 Why is there no noticeable difference in the performance of +/-MergeStores

What did you do to find out yourself? Did you use the trace flags to see if there is a difference in what is optimized / the output assembly code?

**Comment by wenshao** [comment]
After jvm_args is set correctly, +/-MergeStores has a significant performance difference.

## List of tests with significant performance improvements
The putBytes series and LittleEndian-based tests show that +MergeStores has a significant performance improvement under x64.
|   | -MergeStores | +MergeStores | delta|
| --- | --- | --- | --- |
| MergeStoreBench.putBytes4 | 4475.891 | 929.327 | 381.63% |
| MergeStoreBench.putBytes4U | 4479.133 | 928.502 | 382.40% |
| MergeStoreBench.putBytes4X | 4477.133 | 929.183 | 381.84% |
| MergeStoreBench.putChars4B | 9008.350 | 5638.550 | 59.76% |
| MergeStoreBench.putChars4BU | 8961.671 | 1144.479 | 683.03% |
| MergeStoreBench.putChars4C | 4485.308 | 1133.473 | 295.71% |
| MergeStoreBench.putChars4L | 9013.570 | 5640.893 | 59.79% |
| MergeStoreBench.putChars4LU | 8957.625 | 1142.796 | 683.83% |
| MergeStoreBench.putChars4LV | 4488.698 | 1134.303 | 295.72% |
| MergeStoreBench.putChars4S | 4485.836 | 1133.430 | 295.78% |
| MergeStoreBench.setIntL | 15430.183 | 2113.544 | 630.06% |
| MergeStoreBench.setIntLU | 17361.730 | 4783.040 | 262.99% |
| MergeStoreBench.setIntRL | 16525.068 | 3244.126 | 409.38% |
| MergeStoreBench.setIntRLU | 14401.071 | 5930.149 | 142.85% |
| MergeStoreBench.setLongL | 31353.713 | 5405.189 | 480.07% |
| MergeStoreBench.setLongLU | 26113.756 | 4287.166 | 509.11% |
| MergeStoreBench.setLongRL | 27232.898 | 4523.658 | 502.01% |
| MergeStoreBench.setLongRLU | 26196.973 | 4798.177 | 445.98% |
| MergeStoreBench.setLongU | 4500.659 | 4271.225 | 5.37% |

## List of tests that were not improved
On x64 machines, all tests of get operations have no significant improvement, and BigEndian put performance has not improved. It is also very common to use BigEndian byte content on little-endian machines. For example, most network protocols are big-endian. It is expected that MergeStore can be supported.
|   | -MergeStores | +MergeStores | delta|
| --- | --- | --- | --- |
| MergeStoreBench.getCharB | 5908.544 | 5903.216 | 0.09% |
| MergeStoreBench.getCharBU | 4853.054 | 4861.850 | -0.18% |
| MergeStoreBench.getCharBV | 3080.971 | 3081.138 | -0.01% |
| MergeStoreBench.getCharC | 2235.832 | 2235.306 | 0.02% |
| MergeStoreBench.getCharL | 6046.201 | 6034.378 | 0.20% |
| MergeStoreBench.getCharLU | 4934.757 | 4494.743 | 9.79% |
| MergeStoreBench.getCharLV | 2221.754 | 2222.086 | -0.01% |
| MergeStoreBench.getIntB | 8002.830 | 8008.578 | -0.07% |
| MergeStoreBench.getIntBU | 9054.151 | 9048.937 | 0.06% |
| MergeStoreBench.getIntBV | 308.274 | 308.438 | -0.05% |
| MergeStoreBench.getIntL | 7885.680 | 7875.204 | 0.13% |
| MergeStoreBench.getIntLU | 8863.323 | 8866.561 | -0.04% |
| MergeStoreBench.getIntLV | 2228.348 | 2228.067 | 0.01% |
| MergeStoreBench.getIntRB | 8636.679 | 8633.762 | 0.03% |
| MergeStoreBench.getIntRBU | 11102.938 | 11105.491 | -0.02% |
| MergeStoreBench.getIntRL | 8975.416 | 8962.822 | 0.14% |
| MergeStoreBench.getIntRLU | 9249.430 | 9258.589 | -0.10% |
| MergeStoreBench.getIntRU | 2510.359 | 2505.505 | 0.19% |
| MergeStoreBench.getIntU | 2493.932 | 2494.808 | -0.04% |
| MergeStoreBench.getLongB | 24811.283 | 24804.034 | 0.03% |
| MergeStoreBench.getLongBU | 14024.209 | 14013.247 | 0.08% |
| MergeStoreBench.getLongBV | 601.852 | 602.426 | -0.10% |
| MergeStoreBench.getLongL | 25073.219 | 25115.247 | -0.17% |
| MergeStoreBench.getLongLU | 14483.618 | 14497.662 | -0.10% |
| MergeStoreBench.getLongLV | 2225.597 | 2225.810 | -0.01% |
| MergeStoreBench.getLongRB | 24832.411 | 24801.799 | 0.12% |
| MergeStoreBench.getLongRBU | 14027.084 | 14026.284 | 0.01% |
| MergeStoreBench.getLongRL | 25008.679 | 25113.927 | -0.42% |
| MergeStoreBench.getLongRLU | 14425.883 | 14493.830 | -0.47% |
| MergeStoreBench.getLongRU | 3059.614 | 3058.726 | 0.03% |
| MergeStoreBench.getLongU | 3049.682 | 3048.266 | 0.05% |
| MergeStoreBench.putBytes4GetBytes | 5880.164 | 5883.995 | -0.07% |
| MergeStoreBench.putChars4BV | 4488.270 | 4486.457 | 0.04% |
| MergeStoreBench.setCharBS | 6088.826 | 6085.857 | 0.05% |
| MergeStoreBench.setCharBV | 3596.210 | 3595.236 | 0.03% |
| MergeStoreBench.setCharC | 4519.981 | 4471.174 | 1.09% |
| MergeStoreBench.setCharLS | 5619.414 | 5618.239 | 0.02% |
| MergeStoreBench.setCharLV | 2248.493 | 2245.939 | 0.11% |
| MergeStoreBench.setIntB | 8039.705 | 8045.113 | -0.07% |
| MergeStoreBench.setIntBU | 17884.223 | 17764.347 | 0.67% |
| MergeStoreBench.setIntBV | 3239.985 | 3227.997 | 0.37% |
| MergeStoreBench.setIntLV | 2128.975 | 2126.187 | 0.13% |
| MergeStoreBench.setIntRB | 13786.186 | 13815.759 | -0.21% |
| MergeStoreBench.setIntRBU | 14747.463 | 14771.017 | -0.16% |
| MergeStoreBench.setIntRU | 5898.169 | 5875.589 | 0.38% |
| MergeStoreBench.setIntU | 4805.170 | 4784.162 | 0.44% |
| MergeStoreBench.setLongB | 31674.058 | 31662.483 | 0.04% |
| MergeStoreBench.setLongBU | 25696.702 | 25674.394 | 0.09% |
| MergeStoreBench.setLongBV | 2168.387 | 2165.313 | 0.14% |
| MergeStoreBench.setLongLV | 2048.737 | 2116.054 | -3.18% |
| MergeStoreBench.setLongRB | 29901.778 | 29909.501 | -0.03% |
| MergeStoreBench.setLongRBU | 24945.914 | 25005.171 | -0.24% |
| MergeStoreBench.setLongRU | 4797.817 | 4795.018 | 0.06% |


## Full tests
|   | -MergeStores | +MergeStores | delta|
| --- | --- | --- | --- |
| MergeStoreBench.getCharB | 5908.544 | 5903.216 | 0.09% |
| MergeStoreBench.getCharBU | 4853.054 | 4861.850 | -0.18% |
| MergeStoreBench.getCharBV | 3080.971 | 3081.138 | -0.01% |
| MergeStoreBench.getCharC | 2235.832 | 2235.306 | 0.02% |
| MergeStoreBench.getCharL | 6046.201 | 6034.378 | 0.20% |
| MergeStoreBench.getCharLU | 4934.757 | 4494.743 | 9.79% |
| MergeStoreBench.getCharLV | 2221.754 | 2222.086 | -0.01% |
| MergeStoreBench.getIntB | 8002.830 | 8008.578 | -0.07% |
| MergeStoreBench.getIntBU | 9054.151 | 9048.937 | 0.06% |
| MergeStoreBench.getIntBV | 308.274 | 308.438 | -0.05% |
| MergeStoreBench.getIntL | 7885.680 | 7875.204 | 0.13% |
| MergeStoreBench.getIntLU | 8863.323 | 8866.561 | -0.04% |
| MergeStoreBench.getIntLV | 2228.348 | 2228.067 | 0.01% |
| MergeStoreBench.getIntRB | 8636.679 | 8633.762 | 0.03% |
| MergeStoreBench.getIntRBU | 11102.938 | 11105.491 | -0.02% |
| MergeStoreBench.getIntRL | 8975.416 | 8962.822 | 0.14% |
| MergeStoreBench.getIntRLU | 9249.430 | 9258.589 | -0.10% |
| MergeStoreBench.getIntRU | 2510.359 | 2505.505 | 0.19% |
| MergeStoreBench.getIntU | 2493.932 | 2494.808 | -0.04% |
| MergeStoreBench.getLongB | 24811.283 | 24804.034 | 0.03% |
| MergeStoreBench.getLongBU | 14024.209 | 14013.247 | 0.08% |
| MergeStoreBench.getLongBV | 601.852 | 602.426 | -0.10% |
| MergeStoreBench.getLongL | 25073.219 | 25115.247 | -0.17% |
| MergeStoreBench.getLongLU | 14483.618 | 14497.662 | -0.10% |
| MergeStoreBench.getLongLV | 2225.597 | 2225.810 | -0.01% |
| MergeStoreBench.getLongRB | 24832.411 | 24801.799 | 0.12% |
| MergeStoreBench.getLongRBU | 14027.084 | 14026.284 | 0.01% |
| MergeStoreBench.getLongRL | 25008.679 | 25113.927 | -0.42% |
| MergeStoreBench.getLongRLU | 14425.883 | 14493.830 | -0.47% |
| MergeStoreBench.getLongRU | 3059.614 | 3058.726 | 0.03% |
| MergeStoreBench.getLongU | 3049.682 | 3048.266 | 0.05% |
| MergeStoreBench.putBytes4 | 4475.891 | 929.327 | 381.63% |
| MergeStoreBench.putBytes4GetBytes | 5880.164 | 5883.995 | -0.07% |
| MergeStoreBench.putBytes4U | 4479.133 | 928.502 | 382.40% |
| MergeStoreBench.putBytes4X | 4477.133 | 929.183 | 381.84% |
| MergeStoreBench.putChars4B | 9008.350 | 5638.550 | 59.76% |
| MergeStoreBench.putChars4BU | 8961.671 | 1144.479 | 683.03% |
| MergeStoreBench.putChars4BV | 4488.270 | 4486.457 | 0.04% |
| MergeStoreBench.putChars4C | 4485.308 | 1133.473 | 295.71% |
| MergeStoreBench.putChars4L | 9013.570 | 5640.893 | 59.79% |
| MergeStoreBench.putChars4LU | 8957.625 | 1142.796 | 683.83% |
| MergeStoreBench.putChars4LV | 4488.698 | 1134.303 | 295.72% |
| MergeStoreBench.putChars4S | 4485.836 | 1133.430 | 295.78% |
| MergeStoreBench.setCharBS | 6088.826 | 6085.857 | 0.05% |
| MergeStoreBench.setCharBV | 3596.210 | 3595.236 | 0.03% |
| MergeStoreBench.setCharC | 4519.981 | 4471.174 | 1.09% |
| MergeStoreBench.setCharLS | 5619.414 | 5618.239 | 0.02% |
| MergeStoreBench.setCharLV | 2248.493 | 2245.939 | 0.11% |
| MergeStoreBench.setIntB | 8039.705 | 8045.113 | -0.07% |
| MergeStoreBench.setIntBU | 17884.223 | 17764.347 | 0.67% |
| MergeStoreBench.setIntBV | 3239.985 | 3227.997 | 0.37% |
| MergeStoreBench.setIntL | 15430.183 | 2113.544 | 630.06% |
| MergeStoreBench.setIntLU | 17361.730 | 4783.040 | 262.99% |
| MergeStoreBench.setIntLV | 2128.975 | 2126.187 | 0.13% |
| MergeStoreBench.setIntRB | 13786.186 | 13815.759 | -0.21% |
| MergeStoreBench.setIntRBU | 14747.463 | 14771.017 | -0.16% |
| MergeStoreBench.setIntRL | 16525.068 | 3244.126 | 409.38% |
| MergeStoreBench.setIntRLU | 14401.071 | 5930.149 | 142.85% |
| MergeStoreBench.setIntRU | 5898.169 | 5875.589 | 0.38% |
| MergeStoreBench.setIntU | 4805.170 | 4784.162 | 0.44% |
| MergeStoreBench.setLongB | 31674.058 | 31662.483 | 0.04% |
| MergeStoreBench.setLongBU | 25696.702 | 25674.394 | 0.09% |
| MergeStoreBench.setLongBV | 2168.387 | 2165.313 | 0.14% |
| MergeStoreBench.setLongL | 31353.713 | 5405.189 | 480.07% |
| MergeStoreBench.setLongLU | 26113.756 | 4287.166 | 509.11% |
| MergeStoreBench.setLongLV | 2048.737 | 2116.054 | -3.18% |
| MergeStoreBench.setLongRB | 29901.778 | 29909.501 | -0.03% |
| MergeStoreBench.setLongRBU | 24945.914 | 25005.171 | -0.24% |
| MergeStoreBench.setLongRL | 27232.898 | 4523.658 | 502.01% |
| MergeStoreBench.setLongRLU | 26196.973 | 4798.177 | 445.98% |
| MergeStoreBench.setLongRU | 4797.817 | 4795.018 | 0.06% |
| MergeStoreBench.setLongU | 4500.659 | 4271.225 | 5.37% |

**Comment by eme64** [comment]
@wenshao I'll look at your results later.

> @eme64 Are there plans to support MergeLoad, and big-endian MergeStore on little-endian machines?

These are all good ideas, and I already discussed it offline with @cl4es . I have lots of tasks I'm working on, and this is on the lowest tier of priorities for me personally. But if someone else wants to jump on that, then I can coach and review.

We could also be interested in "MergeCopy", i.e. load->store patterns. Maybe this just ends up being SuperWord again, but this time for straight line code.

**Comment by eme64** [comment]
@wenshao Ah. I only just realized it: you have a lot of `get` benchmarks... they don't really belong to `MergeStores`... if anything you could put them in a separate `MergeLoads` benchmark!

Also: now we have lots of data here. But data alone is kind of pointless. We need analysis to see **what patterns** and **why** they get speedups.

**Comment by wenshao** [comment]
@eme64 I have updated the benchmark numbers above, removing the getXXX part. I have also added some analysis. I found that the test using VarHandle is the fastest, whether it is BigEndian or LittleEndian. All the BigEndian tests are not MergeStored, including the Reverse combination scenario.

**Review by eme64 — COMMENTED** [review]
Hi @wenshao ! I'm sorry I lost track of this one.

It looks quite reasonable, though I was wondering why in one benchmark you increment `i` and use `i * 4` as offset, and in the other you decrement `i` and use `off += 4` as offset.

File: test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java [hunk 2]
```diff
@@ -211,613 +99,396 @@ public void setIntB(Blackhole BH) {
 
     @Benchmark
     public void setIntBU(Blackhole BH) {
-        int sum = 0;
-        for (int i = 0; i < ints.length; i++) {
-            int v = ints[i];
-            setIntBU(bytes4, i * 4, v);
-            sum += v;
+        int off = 0;
+        for (int i = ints.length - 1; i >= 0; i--) {
```
[the remark this answers is not in the document]
Comment (wenshao) [reply test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java:103]: My original intention was to have reads and writes in different orders to avoid being optimized into a batch copy.

File: test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java [hunk 3]
```diff
@@ -211,613 +99,396 @@ public void setIntB(Blackhole BH) {
...
     public void setIntLU(Blackhole BH) {
-        int sum = 0;
-        for (int i = 0; i < ints.length; i++) {
-            int v = ints[i];
-            setIntLU(bytes4, i * 4, v);
-            sum += v;
+        int off = 0;
+        for (int i = ints.length - 1; i >= 0; i--) {
+            setIntLU(bytes4, off, ints[i]);
+            off += 4;
```
[the remark this answers is not in the document]
Comment (wenshao) [reply test/micro/org/openjdk/bench/vm/compiler/MergeStoreBench.java:135]: I guess +=4 has less overhead, so the performance test is more accurate.

**Review by eme64 — APPROVED** [review]
Sorry it took me this long. I think it is ok as it is.

**Comment by wenshao** [comment]
Thanks, Emanuel Peter. This is a low-priority PR, so it's okay if it takes some time to get to it.

**Comment by wenshao** [comment]
> These are all good ideas, and I already discussed it offline with @cl4es . I have lots of tasks I'm working on, and this is on the lowest tier of priorities for me personally. But if someone else wants to jump on that, then I can coach and review.
> 
> We could also be interested in "MergeCopy", i.e. load->store patterns. Maybe this just ends up being SuperWord again, but this time for straight line code.

PR #23030  has been submitted to add support for BigEndian MergeStore
